tangos/live_calculation/query_masking.py

```diff
@@ -1,31 +1,43 @@
 from __future__ import absolute_import
 import numpy as np
-from ..util import is_not_none
+from ..util import is_not_none, is_not_false
 
 
 class QueryMask(object):
     """A helper class to remove rows that do not need to be queried, representing them by "None" in the final results"""
     def __init__(self):
         self.N = None
 
-    def mark_nones_as_masked(self, input):
-        """Mark any rows in the input that are None as masked, excluding them from future queries."""
+    def _retain_row_if_true(self, input, condition):
         if self.N is None:
             self.N = len(input)
             to_mask = input
+            self._mask_array = np.full(self.N,True)
         else:
             to_mask = self.unmask(input)
 
         if len(to_mask.shape)==2:
-            mask = np.any(is_not_none(to_mask),axis=0).reshape(self.N)
+            mask = np.any(condition(to_mask),axis=0).reshape(self.N)
         elif len(to_mask.shape)==1:
-            mask = is_not_none(to_mask)
+            mask = condition(to_mask)
         else:
             raise ValueError("Not able to use an input of this shape to determine a query mask")
 
+        mask = mask.astype(np.bool) & self._mask_array.astype(np.bool)
+
         assert mask.shape==(self.N,)
+        self._mask_array = mask
         self.results_target = np.where(mask)
 
+
+    def mark_nones_as_masked(self, input):
+        """Mark any rows in the input that are None as masked, excluding them from future queries."""
+        self._retain_row_if_true(input, is_not_none)
```

Comment:
Can I make an annoying suggestion that you rename `_retain_row_if_true` to `_retain_row_if`? I think this would be clearer (otherwise the bit below about `is_not_false` becomes v confusing!)